test/infrastructure/docker/internal/docker/machine.go

```diff
@@ -419,7 +423,65 @@ func (m *Machine) SetNodeProviderID(ctx context.Context, c client.Client) error
 	return nil
 }
 
-func (m *Machine) getKubectlNode(ctx context.Context) (*types.Node, error) {
+// CloudProviderNodePatch sets the docker provider ID for the kubernetes node.
+func (m *Machine) CloudProviderNodePatch(ctx context.Context, c client.Client, dockerMachine *infrav1.DockerMachine) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	kubectlNode, err := m.getNode(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "unable to set NodeProviderID. error getting a kubectl node")
+	}
+	if !kubectlNode.IsRunning() {
+		return errors.Wrapf(ContainerNotRunningError{Name: kubectlNode.Name}, "unable to set NodeProviderID")
+	}
+
+	node := &corev1.Node{}
+	if err = c.Get(ctx, apimachinerytypes.NamespacedName{Name: m.ContainerName()}, node); err != nil {
+		return errors.Wrap(err, "failed to retrieve node")
+	}
+
+	patchHelper, err := patch.NewHelper(node, c)
+	if err != nil {
+		return err
+	}
+
+	// Set the providerID on the node.
+	log.Info("Setting Kubernetes node providerID")
+	node.Spec.ProviderID = m.ProviderID()
+
+	// If the node is managed by an external cloud provider - e.g. in dualstack tests - add the
+	// machine addresses on the node.
+	if hasCloudProviderTaint(node) {
+		nodeAddressMap := map[corev1.NodeAddress]bool{}
+		for _, addr := range node.Status.Addresses {
```

Comment:
I assume this logic is all about appending addresses and thus not changing the order. Let's please add a godoc comment with the high-level background information why we do this